File: trunk/src/test/java/net/esper/eql/core/TestSelectExprProcessorFactory.java
package net.esper.eql.core;

import java.util.LinkedList;
import java.util.List;

import junit.framework.TestCase;
import net.esper.support.eql.SupportExprNodeFactory;
import net.esper.support.eql.SupportStreamTypeSvc3Stream;
import net.esper.support.event.SupportEventAdapterService;
import net.esper.eql.spec.SelectExprElementNamedSpec;
import net.esper.eql.expression.ExprNode;
import net.esper.eql.expression.ExprValidationException;

public class TestSelectExprProcessorFactory extends TestCase
{
    public void testGetProcessorInvalid() throws Exception
    {
        List<SelectExprElementNamedSpec> selectionList = new LinkedList<SelectExprElementNamedSpec>();
        ExprNode identNode = SupportExprNodeFactory.makeIdentNode("doubleBoxed", "s0");
        ExprNode mathNode = SupportExprNodeFactory.makeMathNode();
        selectionList.add(new SelectExprElementNamedSpec(identNode, "result"));
        selectionList.add(new SelectExprElementNamedSpec(mathNode, "result"));

        try
        {
            SelectExprProcessorFactory.getProcessor(selectionList, null, new SupportStreamTypeSvc3Stream(),
                    null);
            fail();
        }
        catch (ExprValidationException ex)
        {
            // Expected
        }
    }

    public void testGetProcessorWildcard() throws Exception
    {
        List<SelectExprElementNamedSpec> selectionList = new LinkedList<SelectExprElementNamedSpec>();
        SelectExprProcessor processor = SelectExprProcessorFactory.getProcessor(selectionList, null, new SupportStreamTypeSvc3Stream(),
                SupportEventAdapterService.getService());
        assertTrue(processor instanceof SelectExprJoinWildcardProcessor);
    }

    public void testGetProcessorValid() throws Exception
    {
        List<SelectExprElementNamedSpec> selectionList = new LinkedList<SelectExprElementNamedSpec>();
        ExprNode identNode = SupportExprNodeFactory.makeIdentNode("doubleBoxed", "s0");
        selectionList.add(new SelectExprElementNamedSpec(identNode, "result"));
        SelectExprProcessor processor = SelectExprProcessorFactory.getProcessor(selectionList, null, new SupportStreamTypeSvc3Stream(),
                SupportEventAdapterService.getService());
        assertTrue(processor != null);
    }

    public void testVerifyNameUniqueness() throws Exception
    {
        // try valid case
        List<SelectExprElementNamedSpec> elements = new LinkedList<SelectExprElementNamedSpec>();
        elements.add(new SelectExprElementNamedSpec(null, "xx"));
        elements.add(new SelectExprElementNamedSpec(null, "yy"));

        SelectExprProcessorFactory.verifyNameUniqueness(elements);

        // try invalid case
        elements.add(new SelectExprElementNamedSpec(null, "yy"));
        try
        {
            SelectExprProcessorFactory.verifyNameUniqueness(elements);
            fail();
        }
        catch (ExprValidationException ex)
        {
            // expected
        }
    }
}

File: NEsper.Tests/eql/core/TestSelectExprProcessorFactory.cs
///////////////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2007 Esper Team. All rights reserved.                                /
// http://esper.codehaus.org                                                          /
// ---------------------------------------------------------------------------------- /
// The software in this package is published under the terms of the GPL license       /
// a copy of which has been included with this distribution in the license.txt file.  /
///////////////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;

using NUnit.Framework;

using net.esper.eql.expression;
using net.esper.eql.spec;
using net.esper.support.eql;
using net.esper.support.events;

namespace net.esper.eql.core
{
	[TestFixture]
	public class TestSelectExprProcessorFactory
	{
	    [Test]
	    public void TestGetProcessorInvalid()
	    {
	        IList<SelectExprElementCompiledSpec> selectionList = new List<SelectExprElementCompiledSpec>();
	        ExprNode identNode = SupportExprNodeFactory.MakeIdentNode("doubleBoxed", "s0");
	        ExprNode mathNode = SupportExprNodeFactory.MakeMathNode();
	        selectionList.Add(new SelectExprElementCompiledSpec(identNode, "result"));
	        selectionList.Add(new SelectExprElementCompiledSpec(mathNode, "result"));

	        try
	        {
	            SelectExprProcessorFactory.GetProcessor(selectionList, false, null,
	                    new SupportStreamTypeSvc3Stream(), null);
	            Assert.Fail();
	        }
	        catch (ExprValidationException ex)
	        {
	            // Expected
	        }
	    }

	    [Test]
	    public void TestGetProcessorWildcard()
	    {
	        IList<SelectExprElementCompiledSpec> selectionList = new List<SelectExprElementCompiledSpec>();
	        SelectExprProcessor processor = SelectExprProcessorFactory.GetProcessor(selectionList, false, null,
	                new SupportStreamTypeSvc3Stream(), SupportEventAdapterService.GetService());
	        Assert.IsTrue(processor is SelectExprJoinWildcardProcessor);
	    }

	    [Test]
	    public void TestGetProcessorValid()
	    {
	        IList<SelectExprElementCompiledSpec> selectionList = new List<SelectExprElementCompiledSpec>();
	        ExprNode identNode = SupportExprNodeFactory.MakeIdentNode("doubleBoxed", "s0");
	        selectionList.Add(new SelectExprElementCompiledSpec(identNode, "result"));
	        SelectExprProcessor processor = SelectExprProcessorFactory.GetProcessor(selectionList, false, null,
	                new SupportStreamTypeSvc3Stream(), SupportEventAdapterService.GetService());
	        Assert.IsTrue(processor != null);
	    }

	    [Test]
	    public void TestVerifyNameUniqueness()
	    {
	        // try valid case
	        IList<SelectExprElementCompiledSpec> elements = new List<SelectExprElementCompiledSpec>();
	        elements.Add(new SelectExprElementCompiledSpec(null, "xx"));
	        elements.Add(new SelectExprElementCompiledSpec(null, "yy"));

	        SelectExprProcessorFactory.VerifyNameUniqueness(elements);

	        // try invalid case
	        elements.Add(new SelectExprElementCompiledSpec(null, "yy"));
	        try
	        {
	            SelectExprProcessorFactory.VerifyNameUniqueness(elements);
	            Assert.Fail();
	        }
	        catch (ExprValidationException ex)
	        {
	            // expected
	        }
	    }
	}
} // End of namespace
